### PR TITLE
Initial Zephyr integration

### PR DESCRIPTION
**Pull request by mniestroj**
TODOs:
- [x] clean up several remaining TODO items in the code-
- [x] basic documentation
- [x] CI build tests or @szczys confirming it works

### CODE REVIEW

File: src/golioth_coap_client.c [hunk 1]
```diff
@@ -239,6 +239,7 @@ static void nack_handler(
     }
 }
 
+#ifndef GOLIOTH_DONT_OVERRIDE_LIBCOAP_LOG_HANDLER
```
Comment (ncmiller) [inline src/golioth_coap_client.c:242]: Is there a reason to use "double negative" here? In my head, I have to translate "if not not override" to "override", so it might be easier to do something like:

```c
#ifdef GOLIOTH_OVERRIDE_LIBCOAP_LOG_HANDLER
```

File: src/include/golioth_debug.h [hunk 2]
```diff
@@ -4,6 +4,11 @@
 #include "golioth_time.h"
 #include <stdbool.h>
 
+#ifndef LOG_TAG_DEFINE
+#define LOG_TAG_DEFINE(tag)                     \
+    static const char *TAG = #tag
```
Comment (ncmiller) [inline src/include/golioth_debug.h:9]: Seems okay. Since this is a breaking change, let's go ahead and add it to the Changelog for the next release.

File: src/golioth_coap_client.c (hunk 1 above)
Comment (mniestroj) [reply src/golioth_coap_client.c]: Updated.

**Review by szczys — APPROVED** [review]
LGTM!

File: examples/zephyr/README.md [hunk 3]
```diff
@@ -0,0 +1,34 @@
+# Golioth Zephyr Examples
+
+This directory contains a set of examples intended to build and run with the latest release of
+Zephyr.
+
+### Install Zephyr SDK
+
+Follow [Zephyr Geting Started Guide](https://docs.zephyrproject.org/latest/develop/getting_started/).
```
Comment (szczys) [inline examples/zephyr/README.md:8]:
```suggestion
Follow [Zephyr Getting Started Guide](https://docs.zephyrproject.org/latest/develop/getting_started/).
```

File: examples/zephyr/README.md [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+
+### Setup west project
+
+``` sh
+cd ~/golioth-workspace
+west init -m https://github.com/golioth/golioth-firmware-sdk.git --mf west-zephyr.yml
+west update
+(cd modules/lib/golioth-firmware-sdk && git submodule update --init --recursive)
```
Comment (szczys) [inline examples/zephyr/README.md:16]:
```suggestion
cd modules/lib/golioth-firmware-sdk && git submodule update --init --recursive
```

File: examples/zephyr/README.md [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+```
+
+### Configure PSK and PSK-ID credentials
+
+Add credentials to ``qemu_x86
```
Comment (szczys) [inline examples/zephyr/README.md:21]:
```suggestion
Add credentials to ``~/golioth-workspace/
```

File: examples/zephyr/README.md [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+### Install Zephyr SDK
+
+Follow [Zephyr Geting Started Guide](https://docs.zephyrproject.org/latest/develop/getting_started/).
+
+### Setup west project
+
+``` sh
+cd ~/golioth-workspace
+west init -m https://github.com/golioth/golioth-firmware-sdk.git --mf west-zephyr.yml
+west update
```
Comment (szczys) [inline examples/zephyr/README.md:15]:
```suggestion
west update
pip install -r zephyr/scripts/requirements.txt
```

**Review by szczys — APPROVED** [review]
@mniestroj, here are some README suggestions that might be helpful.

File: examples/zephyr/README.md (hunk 3 above)
Comment (mniestroj) [reply examples/zephyr/README.md]: Done.

File: examples/zephyr/README.md (hunk 6 above)
Comment (mniestroj) [reply examples/zephyr/README.md:15]: Two issues I have with this:
- this additional step does not need to be invoked every time and it is a general step in order to prepare for Zephyr development (which should be done as part of `### Install Zephyr SDK` section)
- some systems might require calling `pip3` instead od `pip` I think, depends on distro

Comment (szczys) [reply examples/zephyr/README.md:15]: I like to include this command because if you don't do it, you get an `elftools` error that doesn't offer the user a way to figure out what is wrong:

```
[5/377] Generating include/generated/kobj-types-enum.h, include/generated/otype-to-str.h, include/generated/otype-to-size.h
FAILED: zephyr/include/generated/kobj-types-enum.h zephyr/include/generated/otype-to-str.h zephyr/include/generated/otype-to-size.h /home/mike/golioth-compile/cross-platform/modules/lib/golioth-firmware-sdk/examples/zephyr/golioth_basics/build/zephyr/include/generated/kobj-types-enum.h /home/mike/golioth-compile/cross-platform/modules/lib/golioth-firmware-sdk/examples/zephyr/golioth_basics/build/zephyr/include/generated/otype-to-str.h /home/mike/golioth-compile/cross-platform/modules/lib/golioth-firmware-sdk/examples/zephyr/golioth_basics/build/zephyr/include/generated/otype-to-size.h 
cd /home/mike/golioth-compile/cross-platform/modules/lib/golioth-firmware-sdk/examples/zephyr/golioth_basics/build/zephyr && /home/mike/golioth-compile/cross-platform/.venv/bin/python3.10 /home/mike/golioth-compile/cross-platform/zephyr/scripts/build/gen_kobject_list.py --kobj-types-output /home/mike/golioth-compile/cross-platform/modules/lib/golioth-firmware-sdk/examples/zephyr/golioth_basics/build/zephyr/include/generated/kobj-types-enum.h --kobj-otype-output /home/mike/golioth-compile/cross-platform/modules/lib/golioth-firmware-sdk/examples/zephyr/golioth_basics/build/zephyr/include/generated/otype-to-str.h --kobj-size-output /home/mike/golioth-compile/cross-platform/modules/lib/golioth-firmware-sdk/examples/zephyr/golioth_basics/build/zephyr/include/generated/otype-to-size.h --include-subsystem-list /home/mike/golioth-compile/cross-platform/modules/lib/golioth-firmware-sdk/examples/zephyr/golioth_basics/build/zephyr/misc/generated/struct_tags.json
Traceback (most recent call last):
  File "/home/mike/golioth-compile/cross-platform/zephyr/scripts/build/gen_kobject_list.py", line 62, in <module>
    import elftools
ModuleNotFoundError: No module named 'elftools'
[6/377] Generating include/generated/driver-validation.h
FAILED: zephyr/include/generated/driver-validation.h /home/mike/golioth-compile/cross-platform/modules/lib/golioth-firmware-sdk/examples/zephyr/golioth_basics/build/zephyr/include/generated/driver-validation.h 
cd /home/mike/golioth-compile/cross-platform/modules/lib/golioth-firmware-sdk/examples/zephyr/golioth_basics/build/zephyr && /home/mike/golioth-compile/cross-platform/.venv/bin/python3.10 /home/mike/golioth-compile/cross-platform/zephyr/scripts/build/gen_kobject_list.py --validation-output /home/mike/golioth-compile/cross-platform/modules/lib/golioth-firmware-sdk/examples/zephyr/golioth_basics/build/zephyr/include/generated/driver-validation.h --include-subsystem-list /home/mike/golioth-compile/cross-platform/modules/lib/golioth-firmware-sdk/examples/zephyr/golioth_basics/build/zephyr/misc/generated/struct_tags.json
Traceback (most recent call last):
  File "/home/mike/golioth-compile/cross-platform/zephyr/scripts/build/gen_kobject_list.py", line 62, in <module>
    import elftools
ModuleNotFoundError: No module named 'elftools'
[7/377] Generating include/generated/syscall_dispatch.c, include/generated/syscall_list.h
ninja: build stopped: subcommand failed.
FATAL ERROR: command exited with status 1: /usr/local/bin/cmake --build /home/mike/golioth-compile/cross-platform/modules/lib/golioth-firmware-sdk/examples/zephyr/golioth_basics/build
```

* yes, it is an additional step that only needs to be run once, but there is no problem if run multiple times
* I don't know the answer to the `pip3` v `pip` question. Our docs say that if you're using a virtualenv (which we recommend) then `pip` will use python3 by default. But my suggestion doesn't hold the user's hand through venv installation

So I defer to you on this one. Eventually we'll have a docs page that walks the user through this process and that will include venv, etc.

Comment (mniestroj) [reply examples/zephyr/README.md:15]: Actually it does say that:
```
    import elftools
ModuleNotFoundError: No module named 'elftools'
```
which makes me automatically run `pip install elftools`. But for a beginner, this is not obvious I can imagine. We don't show how to install Zephyr SDK here as well. So either all or nothing. I'd like to keep this document minimal, so that for Zephyr aware users there is just enough documentation on how to use this SDK as part of larger project, or alternatively how to use this SDK to try out the examples in this SDK. There are of course users that are new to Zephyr and this is why I prefer to reference Zephyr Getting Started Guide above. This will be **always** the most up to date documentation on how to start with Zephyr. It also makes clear that "this is a component of Zephyr, so in order to learn Zephyr go to their website". It also offloads us from maintaining instructions on how to setup Zephyr.

TLDR: I think it should be all or nothing (or minimal strictly speaking) on this page and I prefer the latter.

> Eventually we'll have a docs page that walks the user through this process and that will include venv, etc.

It makes sense to have those, I understand. Those will be "all in one place" kind of documentation with all the required steps. I think it will make sense to put a link to such page from this README file with a title "Step by step guide" or something like that. So that way:
- Zephyr users will not be bothered with all the boilerplate documentation about using Zephyr
- new users that see a "half a screen" page (relatively short page) will quickly look for links and should easily spot "Step by step guide".
